method/SpaceFlow/method_spaceflow.py

```diff
@@ -0,0 +1,169 @@
+parser.add_argument(
+    "--dim_red",
+    help="Reduced dimensionality representation (e.g. PCA).",
+    required=False,
+)
+parser.add_argument("--image", help="Path to H&E staining.", required=False)
+parser.add_argument(
+    "--n_clusters", help="Number of clusters to return.", required=True, type=int
+)
+parser.add_argument(
+    "--technology",
+    help="The technology of the dataset (Visium, ST, imaging-based).",
+    required=True,
+)
+parser.add_argument(
+    "--seed", help="Seed to use for random operations.", required=True, type=int
+)
+parser.add_argument(
+    "--config",
+    help="Optional config file (json) used to pass additional parameters.",
+    required=False,
+)
+
+args = parser.parse_args()
+
+from pathlib import Path
+
+out_dir = Path(args.out_dir)
+
+# Output files
+label_file = out_dir / "domains.tsv"
+embedding_file = out_dir / "embedding.tsv"
+
+n_clusters = args.n_clusters
+technology = args.technology
+seed = args.seed
+
+# Load config file
+import json
+
+with open(args.config, "r") as f:
+    config = json.load(f)
+    
+def get_anndata(args):
+    import anndata as ad
+    import numpy as np
+    import pandas as pd
+    from scipy.io import mmread
+    from scipy.sparse import issparse
+    from PIL import Image
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(obs=observations, var=features, obsm={"spatial": coordinates})
+
+    if args.matrix is not None:
+        X = mmread(args.matrix)
+        if issparse(X):
+            X = X.tocsr()
+        adata.X = X
+
+    if args.neighbors is not None:
+        adata.obsp["spatial_connectivities"] = sp.io.mmread(args.neighbors).T.tocsr()
+
+    # Filter by selected samples/features
+    if "selected" in adata.obs.columns:
+        adata = adata[observations["selected"].astype(bool), :]
+    if "selected" in adata.var.columns:
+        adata = adata[:, features["selected"].astype(bool)]
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.img))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+
+adata = get_anndata(args)
+
+
+# Set the seed
+import random
+random.seed(seed)
+
+import scanpy as sc
+from SpaceFlow import SpaceFlow
+import pandas as pd
+import warnings
+import torch
+
+use_cuda = torch.cuda.is_available()
+device = 1 if use_cuda else 0
+
+# Filter genes expressed in < 3 cells
+sc.pp.filter_genes(adata, min_cells=3)
+
+# Create a SpaceFlow object and preprocess it
+sf = SpaceFlow.SpaceFlow(adata=adata)
+sf.preprocessing_data(n_top_genes=3000)
```

Comment:
This preprocessing includes quite a few steps. Check if you can preovide preprocessed data already? Then we would take the preprocessing out of this script.

Reply:
I removed the steps up until PCA and implemented the graph construction implicitly. Seems to work ok. Then later processed data should be passed to the method.